ui-admin/src/portal/languages/PortalEnvLanguageEditor.tsx

```diff
@@ -0,0 +1,230 @@
+              value: row.original.languageName,
+              label: row.original.languageName
+            }}
+            onChange={e => e && setNewItem(
+              {
+                ...row.original,
+                languageName: e.value,
+                languageCode: LANGUAGE_OPTIONS.find(lang => lang.languageName === e.value)?.languageCode || '',
+                isEditing: true
+              })}
+          />
+        }
+        return value
+      }
+    }, {
+      header: 'Code',
+      accessorKey: 'languageCode',
+      cell: ({ row }) => {
+        const value = row.original.languageCode
+        if (isEditable(row.original)) {
+          return row.original.isEditing && <input
+            type="text"
+            aria-label={'Language code'}
+            onChange={e => setNewItem({
+              ...row.original,
+              languageName: e.target.value,
+              isEditing: true
+            })}
+            value={row.original.languageCode}
+          />
+        }
+        return value
+      }
+    }]
+    return readonly ? baseCols : baseCols.concat({
+      header: 'Actions',
+      id: 'actions',
+      cell: ({ row }) => {
+        if (isEditable(row.original)) {
+          if (!row.original.isEditing) {
+            return <button
+              className='btn btn-primary border-0'
+              onClick={() => setNewItem({
+                ...makeEmptyNewItem(),
+                isEditing: true
+              })}>
+              <FontAwesomeIcon icon={faPlus} aria-label={'Add New'}/>
+            </button>
+          }
+
+          return <>
+            <Button
+              className='btn btn-success me-2'
+              disabled={
+                isEmpty(row.original.languageName)
+                || isEmpty(row.original.languageCode)}
+              onClick={() => updateItem({
+                languageCode: row.original.languageCode,
+                languageName: row.original.languageName,
+                id: ''
+              })}>
+              <FontAwesomeIcon icon={faCheck} aria-label={'Accept'}/>
+            </Button>
+            <Button className='btn btn-danger' onClick={() => setNewItem(makeEmptyNewItem())}>
+              <FontAwesomeIcon icon={faX} aria-label={'Cancel'}/>
+            </Button>
+          </>
+        }
+        return <Button className='btn btn-outline-danger border-0' onClick={() => {
+          setItemSelectedForDeletion(row.original)
+        }}>
+          <FontAwesomeIcon icon={faTrashCan} aria-label={'Delete'}/>
+        </Button>
+      }
+    })
+  }, [items, newItem, readonly])
+
+  const data = useMemo(
+    () => (items as PortalEnvironmentLanguageRow[]).concat(newItem),
+    [items, newItem])
+
+  const table = useReactTable<PortalEnvironmentLanguageRow>({
+    data,
+    columns,
+    getCoreRowModel: getCoreRowModel()
+  })
+
+  return <div className='px-3 pt-1'>
+    {basicTableLayout(table, { tdClass: 'col-1 ' })}
```

Comment:
Love the answer mapping-style table becoming more common!

From a UX perspective, I feel like the whole table being viewable at the width of the screen dominates the page visually when you scroll down. I wonder if it would be better to have the table only show up if you go into an edit mode on the languages, otherwise it has a more condensed view (and the save languages button disabled or not visible). But I feel like the whole page needs UX love, so I'm cool either way

Reply:
made a change to make the table half-width on big screens so it doesn't dominate as much -- good call.   Agree we can revisit the whole page at a later time